### PR TITLE
Any plan on CUDA support?

### PR DESCRIPTION
**Pull request by panqingfeng**
Here is a straight forward CUDA port, is it possible to merge to fastText master stream?
just clone https://github.com/panqingfeng/fastText.git, make -f Makefile.cuda (make sure cuda is installed), and play with it.

in my case, a job with
Number of words:  4094582
Number of labels: 23770
now cost 8 hours(with a 1080ti 8Gmem GPU), which I need 18 hours on a E5-2420 0 @ 1.90GHz cpu

It may do better if I do some performance tunning on CUDA code.

### CODE REVIEW

**Comment by ruckc** [comment]
+1

**Comment by panqingfeng** [comment]
> +1

My branche is working on supervised training, it seems master branch has a huge interface rewrite after I forked my branch. I will port my code to the new interface soon.

**Comment by panqingfeng** [comment]
> > +1
> 
> My branche is working on supervised training, it seems master branch has a huge interface rewrite after I forked my branch. I will port my code to the new interface soon.

I have finished it. Now support a new loss type "cuda_softmax".
Just checkout the code and build with: make fasttext_cuda
run with fasttext_cuda -loss cuda_softmax $(other args)
It seems 3x to 10x faster than softmax running with cpu
Please enjoy it.

**Comment by Celebio** [comment]
Hi @panqingfeng ,
Nice job! Thank you very much. We will review it. Unfortunately we can't tell when and if we will merge it (it may or may not conflict with our internal roadmap).
Great job anyway. Thank you again.

Onur

**Comment by Liangtaiwan** [comment]
@panqingfeng I try to use this branch today. However, in my experiment. The cuda version is 100 times (cuda_softmax) and 10 times slower(softmax) than original one.
I think the reason is you disable batching in the commit  9dd9e31. As a result, the program do not benefit from  using gpu since it cannot parallelize the calculation.

**Comment by panqingfeng** [comment]
> @panqingfeng I try to use this branch today. However, in my experiment. The cuda version is 100 times (cuda_softmax) and 10 times slower(softmax) than original one.
> I think the reason is you disable batching in the commit [9dd9e31](https://github.com/facebookresearch/fastText/commit/9dd9e31ec5668df6db79e3c8dbed5c786801c14d). As a result, the program do not benefit from using gpu since it cannot parallelize the calculation.

How many output do you have? Like how many tags do you have if you are using supervised learning? If it's less than 100(just a lucky guess) , you shoud keep using CPU. 
How many samples did you put to train? with cuda_softmax, it have to do some memory copy job first, it should take a few seconds.
CUDA is much faster than CPU if you need a lot compute, but the data have to transfer from host(CPU) to device(GPU) before compute and fransfer back after, so if it's a little to compute, it's not worth.
And, I didn't modify the original part( I just inherit the base class, so if you use softmax, you should using the original implementation) , 10 times slower than origianal seems weird.
Yes, I gaveup the batch training because it make the interface not graceful, it lost about 10%-15% performance in my testcase. If I was proven wrong, I can revert to code for performance.

Anyway, let's figure out what happen first, can you feed back more information?

**Comment by Liangtaiwan** [comment]
@panqingfeng Thanks for your reply
The script I used is
https://github.com/facebookresearch/UnsupervisedMT/blob/master/NMT/get_data_enfr.sh
and I slightly modify line 284 by add -loss cuda_softmax and change $FASTTEXT to fasttext_cuda

**Comment by panqingfeng** [comment]
> @panqingfeng Thanks for your reply
> The script I used is
> https://github.com/facebookresearch/UnsupervisedMT/blob/master/NMT/get_data_enfr.sh
> and I slightly modify line 284 by add -loss cuda_softmax and change $FASTTEXT to fasttext_cuda

@Liangtaiwan 
I didn't run this scipt before, but as far as I can tell, the last command in this script matter
if ! [[ -f "$CONCAT_BPE.vec" ]]; then
  echo "Training fastText on $CONCAT_BPE..."
  $FASTTEXT skipgram -epoch $N_EPOCHS -minCount 0 -dim 512 -thread $N_THREADS -ws 5 -neg 10 -input $CONCAT_BPE -output $CONCAT_BPE
fi

As you can see, it use skipgram mode, which use **ns** as default loss function.
1. cuda_softmax is Just a speed up of softmax algorithm
2. ns is much faster than softmax (for example, a classification problem with 10k tags, softmax have to comput all 10k softmax ouput and get the result,  ns just  compute  1 positive and 5 other random negative outputs, ns will aways defeat softmax no mater use GPU or not.)
3. cpu is a better choice if you are usning ns. (there is not much to comput, so transfer data from/to GPU may cost more time)
4. a fair compair should be softmax vs cuda_softmax, can you do it with that samples and get us back the result?
